Framework/Algorithms/src/SortXAxis2.cpp

```diff
@@ -0,0 +1,152 @@
+#include "MantidAlgorithms/SortXAxis2.h"
+
+using namespace Mantid::Kernel;
+using namespace Mantid::API;
+
+namespace Mantid {
+namespace Algorithms {
+
+DECLARE_ALGORITHM(SortXAxis)
+
+const std::string SortXAxis::name() const { return "SortXAxis"; }
+
+int SortXAxis::version() const { return 2; }
+
+const std::string SortXAxis::category() const {
+  return "Transforms\\Axes;Utility\\Sorting";
+}
+
+const std::string SortXAxis::summary() const {
+  return "Clones the input MatrixWorkspace(s) and orders the x-axis in an "
+         "ascending fashion.";
+}
+
+void SortXAxis::init() {
+  declareProperty(make_unique<WorkspaceProperty<MatrixWorkspace>>(
+                      "InputWorkspace", "", Direction::Input),
+                  "Input Workspace");
+
+  declareProperty(make_unique<WorkspaceProperty<MatrixWorkspace>>(
+                      "OutputWorkspace", "", Direction::Output),
+                  "Sorted Output Workspace");
+
+  auto orderingValues = std::vector<std::string>({"Ascending", "Descending"});
+  auto orderingValidator =
+      boost::make_shared<StringListValidator>(orderingValues);
+  declareProperty("Ordering", orderingValues[0], orderingValidator,
+                  "Ascending or descending sorting", Direction::Input);
+}
+
+void SortXAxis::exec() {
+
+  MatrixWorkspace_const_sptr inputWorkspace = getProperty("InputWorkspace");
+  MatrixWorkspace_sptr outputWorkspace = inputWorkspace->clone();
+
+  for (auto specNum = 0u; specNum < inputWorkspace->getNumberHistograms();
+       specNum++) {
+
+    const auto sizeOfX = inputWorkspace->x(specNum).size();
+
+    auto workspaceIndicies = createIndexes(sizeOfX);
+
+    sortIndicesByX(workspaceIndicies, getProperty("Ordering"), inputWorkspace,
+                   specNum);
+
+    copyToOutputWorkspace(workspaceIndicies, inputWorkspace, outputWorkspace,
+                          sizeOfX, inputWorkspace->y(specNum).size(), specNum);
+  }
+  setProperty("OutputWorkspace", outputWorkspace);
+}
+
+std::vector<std::size_t> SortXAxis::createIndexes(const size_t sizeOfX) {
+  std::vector<std::size_t> workspaceIndicies;
+  workspaceIndicies.reserve(sizeOfX);
+  for (auto workspaceIndex = 0u; workspaceIndex < sizeOfX; workspaceIndex++) {
+    workspaceIndicies.emplace_back(workspaceIndex);
+  }
+  return workspaceIndicies;
+}
+
+void SortXAxis::sortIndexesByAscendingXValue(
```

Comment:
`sortIndexesByAscendingXValue` and `sortIndexesByDescendingXValue` contain redundancy (only difference is < or >): I think there is a better way of doing it while not repeating the code.

Reply:
I've struggled to find an exact method (still have not) to what you described but I can just move them both into their parent function. If you have any ideas on how to improve this because it is two completely separate comparators for the sort function.

Reply:
You could use...
```c++
template <typename Comparator>
void sortByXValue(std::vector<std::size_t> &workspaceIndicies,
                  MatrixWorkspace_const_sptr inputWorkspace, 
                  unsigned int specNum,
                  Comparator const& compare) {
  std::sort(workspaceIndicies.begin(), workspaceIndicies.end(), 
    [&](std::size_t lhs, std::size_t rhs) -> bool {
      return compare(inputWorkspace->x(specNum)[lhs], 
                     inputWorkspace->x(specNum)[rhs]);
    });
}
```

which can be called by
```c++
sortByXValue(workspaceIndices, inputWorkspace, specNum, std::less<double>());
```
or 
```c++
sortByXValue(workspaceIndices, inputWorkspace, specNum, std::greater<double>());
```